File: dblp/scala/src/main/scala/com/nephometrics/dblp/CitationsRunner.scala
// The main object for running the citations pipeline.

package com.nephometrics.dblp

import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.types._

/**
 * Example Usage:
 * {{{
 * sbt run CitationsRunner <path to output> <input files>
 * If you have wildcards in the input files, enclose them in quotes
 * to avoid the shell expanding the wildcards.
 * }}}
 */
object CitationsRunner {
  def main(args: Array[String]) {
    if (args.length < 2) {
      System.err.println("Usage: CitationsRunner <output path> <input files>")
      System.exit(1)
    }

    //val filenames = args.slice(1, args.length)
    val filenames = args(1)
    val outputs = args(0)

    System.err.println("outputs to to " + outputs)
    System.err.println("inputs: " + filenames)

    val spark = SparkSession
      .builder
      .appName("citations")
      .getOrCreate()

    //spark.sparkContext.setLogLevel(org.apache.log4j.Level.ERROR.toString())
    val stats = new Citations(spark)

    // You can't initialize this using _ because the JVM doesn't know
    // how to create a default instance. The null.asInstanceOf[T] pattern
    // is apparently the recommended option.
    var baseData:DataFrame = null.asInstanceOf[DataFrame]

    // See if we're loading json or parquet data.
    if (filenames.endsWith("json")) {
      // This selects a subset of fields and specifies types and nullability,
      // which is useful because I know the id field is not null.
      // Also ensures year is an integer not a long.
      val schema = StructType(Array(
        StructField("id", StringType, false),
        StructField("references", ArrayType(StringType, false), true),
        StructField("year", IntegerType, false)))
        baseData = spark.read.schema(schema).json(filenames)
    } else {
      // You can't use a schema to read a parquet file when that requires
      // typecasting (UnsupportedOperationException), so read it like this,
      // the later 'select' statements will be clever enough to only read
      // what they need.
      val tmp = spark.read.parquet(filenames)
      baseData = tmp.withColumn("year", tmp("year").cast(IntegerType))
    }
    val idRefs = stats.idRefYearDs(baseData)
    val counted = stats.countCitationsByYear(idRefs)
    val cited = stats.countCitationsByAge(counted, baseData)
    val history = stats.collectCitationHistory(cited)

    // 10 is based on experience, this will lead to about 20MB per file.
    history.repartition(10).write.format("json").save(outputs)

    spark.stop()
  }
}

File: dblp/python/citationsRunner.py
import os
import getopt
import sys

from pyspark.sql import SparkSession
from pyspark.sql.types import StructType, StructField, StringType, ArrayType, IntegerType

import citationsCommon, citations, citationsDF

# Make sure these are set
#os.environ["Java_HOME"] = ...
#os.environ["SPARK_HOME"] = ...
os.environ["PYTHONHASHSEED"] = "2"

if __name__ == "__main__":

    if len(sys.argv) < 3:
        sys.exit("Usage: citationsRunner <outputdir> <inputfiles>")

    ovpairs, args = getopt.getopt(sys.argv[1:], "d", ["dataframe"])

    useDataFrames = False
    for opt, _ in ovpairs:
        if opt in ['-d', '--dataframe']:
            useDataFrames = True

    outputdir, *inputfiles = args

    print("outputdir: ", outputdir)
    print("inputfiles: ", inputfiles)
    if not inputfiles:
        sys.exit("Need an input file")

    useJson = inputfiles[0].endswith("json")

    spark = SparkSession.builder.getOrCreate()
    spark.conf.set("spark.sql.parquet.binaryAsString", True)
    sc = spark.sparkContext

    citationData = None
    if useJson:
      # prepare the schema
      schema = StructType([
          StructField('id', StringType(), False),
          StructField('references', ArrayType(StringType(), False), True),
          StructField('year', IntegerType(), False),
      ])
      citationData = spark.read.json(inputfiles, schema=schema)
    else:
      # parquet does not support schema
      citationData = spark.read.parquet(inputfiles[0])

    # citCountArrays: id, [array of citation counts by year]
    citCountArrays = None
    if useDataFrames:
      citCountArrays = citationsDF.citationCountsE2E(citationData, 34)
    else:
      citCountArrays = citations.citationCountsE2E(citationData, 34)

    averageByAggregate = citationsCommon.averageAggregates(citCountArrays, 100)

    citCountArrays.coalesce(10).saveAsTextFile(
            '{}/count_arrays.csv'.format(outputdir))
    open('{}/averages'.format(outputdir) , 'w').write(
            ",".join([str(x) for x in averageByAggregate]))
 
    spark.stop()
